moto/firehose/models.py

```diff
@@ -0,0 +1,644 @@
+            # S3Configuration becomes S3DestinationDescription for the
+            # other destinations.
+            self.destinations[0][destination_name][
+                "S3DestinationDescription"
+            ] = destination_config["S3Configuration"]
+            del self.destinations[0][destination_name]["S3Configuration"]
+
+        self.delivery_stream_status = "ACTIVE"
+        self.delivery_stream_arn = f"arn:aws:firehose:{region}:{ACCOUNT_ID}:/delivery_stream/{delivery_stream_name}"
+
+        self.create_timestamp = str(datetime.utcnow())
+        self.version_id = "1"  # Used to track updates of destination configs
+
+        # I believe boto3 only adds this field after an update ...
+        self.last_update_timestamp = str(datetime.utcnow())
+
+
+class FirehoseBackend(BaseBackend):
+    """Implementation of Firehose APIs."""
+
+    def __init__(self, region_name=None):
+        self.region_name = region_name
+        self.delivery_streams = {}
+        self.tagger = TaggingService()
+
+    def lookup_name_from_arn(self, arn):
+        """Given an ARN, return the associated delivery stream name."""
+        # TODO - need to test
+        return self.delivery_streams.get(arn.split("/")[-1])
+
+    def reset(self):
+        """Re-initializes all attributes for this instance."""
+        region_name = self.region_name
+        self.__dict__ = {}
+        self.__init__(region_name)
+
+    def create_delivery_stream(
+        self,
+        region,
+        delivery_stream_name,
+        delivery_stream_type,
+        kinesis_stream_source_configuration,
+        delivery_stream_encryption_configuration_input,
+        s3_destination_configuration,
+        extended_s3_destination_configuration,
+        redshift_destination_configuration,
+        elasticsearch_destination_configuration,
+        splunk_destination_configuration,
+        http_endpoint_destination_configuration,
+        tags,
+    ):  # pylint: disable=too-many-arguments,too-many-locals,unused-argument
+        """Create a Kinesis Data Firehose delivery stream."""
+        (destination_name, destination_config) = find_destination_config_in_args(
+            locals()
+        )
+
+        if delivery_stream_name in self.delivery_streams:
+            raise ResourceInUseException(
+                f"Firehose {delivery_stream_name} under accountId {ACCOUNT_ID} "
+                f"already exists"
+            )
+
+        if len(self.delivery_streams) == DeliveryStream.MAX_STREAMS_PER_REGION:
+            raise LimitExceededException(
+                f"You have already consumed your firehose quota of "
+                f"{DeliveryStream.MAX_STREAMS_PER_REGION} hoses. Firehose "
+                f"names: {list(self.delivery_streams.keys())}"
+            )
+
+        # Rule out situations that are not yet implemented.
+        if delivery_stream_encryption_configuration_input:
+            raise NotImplementedError(
```

Comment:
This should be a warning IMO - we can still continue with the creation, it just will not have the full functionality

Reply:
By warning, what do you mean?  moto doesn't seem to use logging.

Reply:
Sorry, you're right, I was being too quick- I should have clarified. I was talking about the OOTB warning library:

```
import warnings
warnings.warn(...)
``` 

It would be the first instance of Moto using it, afaik!

Reply:
I just checked and there are a couple other places using warnings - cloudformation, ec2, emr, events, elastictranscoder, packages.  I think it might be the first time I've used the warnings package.  

It makes sense to issue a warning for the encryption configuration, but I'm a bit concerned about just issuing a warning for a Splunk destination.  Any attempt to put records to that delivery stream will fail and it would be confusing.

Reply:
There's still a lot of value in testing whatever business logic leads up to the Splunk-connector though - that all becomes impossible if the `create_delivery_stream` throws an Exception.

Personally I wouldn't want to test the Splunk-connector anyway, or push data into Splunk as part of my unit tests and going to the trouble of connecting to it to verify the data has landed. I'd trust that AWS handles that part properly, and all I care about is that the surrounding logic is correct.

Reply:
Ok.  I've changed them to warnings.  Personally I find it frustrating when our Terraform tests fail and I'm not sure what caused the initial error.  But I'm working with a large number of repos that were already tested against AWS and we're now trying to automate those tests using moto.

Reply:
Thanks. Yeah, I can understand it's frustrating. I expect the majority of users to have more straight forward test cases, and for them a warning will be more useful - hence my insistence 🙂